addons/website/models/website_visitor.py

```diff
@@ -0,0 +1,177 @@
+        cookie_content = request.httprequest.cookies.get('visitor_id')
+        if cookie_content and '-' in cookie_content:
+            visitor_id, visitor_hash = cookie_content.split('-', 1)
+            if _consteq(visitor_hash, self._get_visitor_hash(visitor_id)):
+                return visitor.sudo().with_context(active_test=False).search([('id', '=', visitor_id)])  # search to avoid having to call exists()
+        return visitor
+
+    def _handle_webpage_dispatch(self, response, is_webpage=False):
+        """ Called on dispatch. This will create a website.visitor if the http request object
+        is a tracked website page. Only on tracked page to avoid having too much operations done on every page
+        or other http requests.
+        Note: The side effect is that the last_connection_datetime is updated ONLY on tracked pages."""
+        website_page = is_webpage and response.qcontext['main_object']
+        if website_page and website_page.is_tracked:
+            visitor_sudo = self._get_visitor_from_request()
+            if not visitor_sudo:
+                # If visitor does not exist
+                visitor_sudo = self._create_visitor(website_page.id)
+                sign = self._encode(visitor_sudo.id)
+                response.set_cookie('visitor_id', sign)
+            else:
+                # Add page even if already in visitor_page_ids as checks on relations are done in many2many write method
+                vals = {
+                    'last_connection_datetime': datetime.now(),
+                    'visitor_page_ids': [(0, 0, {'page_id': website_page.id, 'visit_date': datetime.now()})],
+                }
+                if visitor_sudo.last_connection_datetime < (datetime.now() - timedelta(hours=8)):
+                    vals['visit_count'] = visitor_sudo.visit_count + 1
+                if not visitor_sudo.active:
+                    vals['active'] = True
+                visitor_sudo.write(vals)
+        return response
+
+    def _create_visitor(self, website_page_id=False):
+        country_code = request.session.geoip and request.session.geoip.get('country_code')
+        country_id = request.env['res.country'].sudo().search([('code', '=', country_code)], limit=1).id
+        lang_id = request.env['res.lang'].sudo().search([('code', '=', request.lang)], limit=1).id
+        vals = {
+            'last_connection_datetime': datetime.now(),
+            'lang_id': lang_id,
+            'country_id': country_id,
+            'website_id': request.website.id
+        }
+        if website_page_id:
+            vals['visitor_page_ids'] = [(0, 0, {'page_id': website_page_id, 'visit_date': datetime.now()})]
+        # Set signed visitor id in cookie
+        visitor_sudo = self.sudo().create(vals)
+        return visitor_sudo
+
+    def _cron_archive_visitors(self):
+        datetime_now = "'" + fields.Datetime.to_string(fields.Datetime.from_string(fields.Datetime.now())) + "'"
```

Comment:
`to_string(from_string()) ??`

Do you really need the 'to_string' part?

Also, as `fields.Datetime.now()` returns something that is understood by the ORM, can't this be:

```
sql = """ UPDATE %s
              SET active = False
              WHERE last_connection_datetime < (%s - INTERVAL '7 days') """ 
self.env.cr.execute(sql, (self._table, fields.Datetime.now()))
```

Or something like that?
It certainly feels over complicated right now.

Reply:
indeed this was complete crap